cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -236,6 +247,138 @@ func canInstallAndRunAtOnce(mst *initramfsMountsState) (bool, error) {
+			defer stdin.Close()
+			_, err = stdin.Write(encoded)
+		}()
+		writeErr <- err
+	}()
+	output, err := cmd.Output()
+	if err != nil {
+		return make([]byte, 0), err
+	}
+	err = <-writeErr
+	if err != nil {
+		return make([]byte, 0), err
+	}
+	return output, nil
+}
+
+func hasFDESetupHook(kernelInfo *snap.Info) (bool, error) {
+	_, ok := kernelInfo.Hooks["fde-setup"]
+	return ok, nil
+}
+
+func doInstall(mst *initramfsMountsState, model *asserts.Model, sysSnaps map[snap.Type]*seed.Snap) error {
+	kernelSnap, err := getSnap(sysSnaps, snap.TypeKernel)
+	if err != nil {
+		return err
+	}
+	baseSnap, err := getSnap(sysSnaps, snap.TypeBase)
+	if err != nil {
+		return err
+	}
+	gadgetSnap, err := getSnap(sysSnaps, snap.TypeGadget)
+	if err != nil {
+		return err
+	}
+	kernelMountDir := filepath.Join(boot.InitramfsRunMntDir, snapTypeToMountDir[snap.TypeKernel])
+	gadgetMountDir := filepath.Join(boot.InitramfsRunMntDir, snapTypeToMountDir[snap.TypeGadget])
+	gadgetInfo, err := gadget.ReadInfo(gadgetMountDir, model)
+	if err != nil {
+		return err
+	}
+	encryptionSupport, err := install.CheckEncryptionSupport(model, secboot.TPMProvisionFull, kernelSnap, gadgetInfo, runFDESetupHook)
+	if err != nil {
+		return err
+	}
+	useEncryption := (encryptionSupport != secboot.EncryptionTypeNone)
+
+	installObserver, trustedInstallObserver, err := install.BuildInstallObserver(model, gadgetMountDir, useEncryption)
+	if err != nil {
+		return err
+	}
+
+	options := gadgetInstall.Options{
+		Mount:          true,
+		EncryptionType: encryptionSupport,
+	}
+	installedSystem, err := gadgetInstallRun(model, gadgetMountDir, kernelMountDir, "", options, installObserver, timings.New(nil))
+	if err != nil {
+		return err
+	}
+
+	if trustedInstallObserver != nil {
+		if err := install.PrepareEncryptedSystemData(model, installedSystem.KeyForRole, trustedInstallObserver); err != nil {
+			return err
+		}
+	}
+
+	err = install.PrepareRunSystemData(model, gadgetMountDir, timings.New(nil))
+	if err != nil {
+		return err
+	}
+
+	bootWith := &boot.BootableSet{
+		Base:                baseSnap,
+		BasePath:            sysSnaps[snap.TypeBase].Path,
+		Gadget:              gadgetSnap,
+		GadgetPath:          sysSnaps[snap.TypeGadget].Path,
+		Kernel:              kernelSnap,
+		KernelPath:          sysSnaps[snap.TypeKernel].Path,
+		UnpackedGadgetDir:   gadgetMountDir,
+		RecoverySystemLabel: mst.recoverySystem,
+	}
+
+	if err := bootMakeRunnableStandaloneSystem(model, bootWith, trustedInstallObserver); err != nil {
+		return err
+	}
+
+	dataMountOpts := &systemdMountOptions{
+		Bind: true,
+	}
+	if err := doSystemdMount(boot.InstallUbuntuDataDir, boot.InitramfsDataDir, dataMountOpts); err != nil {
+		return err
+	}
+
+	currentSeed, err := mst.LoadSeed(mst.recoverySystem)
+	if err != nil {
+		return err
+	}
+	preseedSeed, ok := currentSeed.(seed.PreseedCapable)
+	if ok {
+		if err := installApplyPreseededData(preseedSeed, boot.InitramfsWritableDir(model, false)); err != nil {
```

Comment:
(super nitpick) maybe: `runMode := false` here and use that in `boot.InitramfsWritableDir(model, runMode))` to make clear what the boolean parameter means (I had to look at the InitramfsWritableDir() signature to see what it means).

Reply:
done